### DISCUSSION TITLE
MAM archive in MUC rooms: archive size

### DISCUSSION THREAD

**Discussion by RichardWepnerHadean** [processone/ejabberd#4215] [Questions] [answered]
Hello,

for a project we're attempting to use ejabberd as a server for the global text chat. In order for moderators to be able to moderate the chat, they will need to be able to scroll back in the chat history to inspect old messages. We intend to implement the global chat using MUC rooms (i. e. `mod_muc`) and the message archival using MAM (i. e. `mod_mam`).

However, when I try to retrieve messages from MAM, the amount of messages stored seems to be limited to whatever the current value of `modules.mod_muc.history_size` (default `20`) is. [The documentation for `history_size` belo `mod_muc`](https://docs.ejabberd.im/admin/configuration/modules/#mod_muc) states the following:

> Setting the value to 0 disables the history feature and, as a result, nothing is kept in memory. [...] NOTE: modern XMPP clients rely on Message Archives (XEP-0313), so feel free to disable the history feature if you’re only using modern clients and have mod_mam module loaded.

Since we only use clients that we implemented for this purpose, we can ensure that all message retrieval is done using MAM, however setting `history_size` to `0` causes archive queries to not retrieve any messages.

What would we need to do (be it in the client implementation or in the ejabberd configuration) in order not to have such a limitation in MAM? Would the only option be to set `history_size` to a large enough value (contradicting the documentation)?

Besides that, I noticed that [the current MAM specification](https://xmpp.org/extensions/xep-0313.html) uses the namespace `urn:xmpp:mam:2` consistently, but Ejabberd only recognizes queries with the namespace `urn:xmpp:mam:0`. Does that mean that Ejabberd implements an old version of the specification like [v 0.3](https://xmpp.org/extensions/attic/xep-0313-0.3.html) or [https://xmpp.org/extensions/attic/xep-0313-0.2.html](v 0.2)? (If so: which one?)

**Comment by weiss** [comment, marked as answer]
> Since we only use clients that we implemented for this purpose, we can ensure that all message retrieval is done using MAM, however setting `history_size` to `0` causes archive queries to not retrieve any messages.

That traditional in-memory history is used as fall-back when MAM storage wasn't enabled (by the client or the owner) in the room configuration.  To enable it for new rooms by default:

```
mod_muc:
  default_room_options:
    mam: true
    persistent: true
```

> I noticed that the current MAM specification uses the namespace `urn:xmpp:mam:2` consistently, but Ejabberd only recognizes queries with the namespace `urn:xmpp:mam:0`.

ejabberd supports all MAM revisions up to  `urn:xmpp:mam:2`.  If a `:2` query failed, we'd have to see the actual query/response to comment on that.
